<a href="https://colab.research.google.com/github/Habib0308/University_Final_Year_Project/blob/master/Notebooks/Chicago_Data_Preprocessing_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Objective of this preprocessing is to create encoded dataset for predicting ward**


**Avoid colab clossing**
```
function ClickConnect(){
    document.querySelector("colab-connect-button").click()
    console.log("Clicked on connect button"); 
}
setInterval(ClickConnect,60000)
```

# **Imports**

In [141]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# mounting google drive

In [142]:
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


# **Importing Dataset**

In [143]:
file_path= "/content/gdrive/My Drive/Final Year Project/Dataset/dataset_analyzed.csv"
data= pd.read_csv(file_path) #, nrows= 10000
del data["date_of_occurrence"]
df= data.copy()

# Changing Data type

In [144]:
df["ward"]= df["ward"].astype(object)
df["beat"]= df["beat"].astype(object)
df["hour"]= df["hour"].astype(object)

In [145]:
df.head(2)

block  iucr primary_description  ... month  week_day hour
0   061XX W GRAND AVE  0486             BATTERY  ...   Jun  Thursday    5
1  043XX N CICERO AVE  0560             ASSAULT  ...   Jun  Thursday    5

[2 rows x 15 columns]

In [146]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 232015 entries, 0 to 232014
Data columns (total 15 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   block                  232015 non-null  object 
 1   iucr                   232015 non-null  object 
 2   primary_description    232015 non-null  object 
 3   secondary_description  232015 non-null  object 
 4   location_description   232015 non-null  object 
 5   arrest                 232015 non-null  object 
 6   domestic               232015 non-null  object 
 7   beat                   232015 non-null  object 
 8   ward                   232015 non-null  object 
 9   fbi_cd                 232015 non-null  object 
 10  latitude               232015 non-null  float64
 11  longitude              232015 non-null  float64
 12  month                  232015 non-null  object 
 13  week_day               232015 non-null  object 
 14  hour                   232015 non-nu

# **One hot encoding for features with many variables**
In this i will pick up top most freqquencty occuring labels from each feature and for the rest of the labels if they occur in the data would have their complete row as zero and they will not contain a column of their own


## Counting columns and unique values for each column before encoding

[onehotencoder](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html)

In [147]:
orig_rows, orig_col= df.shape
print(" Rows",orig_rows,"\n","Columns",orig_col)

 Rows 232015 
 Columns 15


In [148]:
for column in df.columns:
    print(column, ": ", df[column].nunique())

block :  26988
iucr :  314
primary_description :  32
secondary_description :  427
location_description :  162
arrest :  2
domestic :  2
beat :  274
ward :  50
fbi_cd :  26
latitude :  115738
longitude :  115721
month :  12
week_day :  7
hour :  24


In [149]:
df.columns

Index(['block', 'iucr', 'primary_description', 'secondary_description',
       'location_description', 'arrest', 'domestic', 'beat', 'ward', 'fbi_cd',
       'latitude', 'longitude', 'month', 'week_day', 'hour'],
      dtype='object')

# **Data For Predicting Ward**

# **Which values to pick?**

**Top Values :**

* top 10 **primary_description**
 * top 10 **secondary_description** for each primary description
 * top 10 **IUDR** codes for each primary description
* top 10 **blocks** from each and all wards
* top 10 **beats**
* top 30 **location** descriptions
* top 10 **FBI code**

**Others**

* arrest :  2
* domestic :  2
* latitude :  115738
* longitude :  115721
* month :  12
* week_day :  7
* hour :  24
 * dividing into 4 times



# **Encoder Function**

This function accepts:
* **column_name**: The column to encode
* **top_values**: List of labels to encode

In [150]:
def encoder_function(column_name, top_values):
    """
    This function accepts:
    column_name: The column to encode
    top_values: List of values to encode
    """
    for category in top_values:
        new_col= column_name + "_"+ category
        df[new_col]= np.where(df[column_name] == category, 1, 0)
    df.head(3)

# **Getting internal top 10 values function**

In [151]:
def list_inner_top_values(column_name, outer_top, no_of_top_topick, Column_to_count):
    """
    This function accepts:
    column_name: The name of the outer level column
    outer_top: (List) top values of outer column
    no_of_top_topick: Number of top values to pick from inner column
    Column_to_count: Name of column from which you want to pick top values of
    """
    top_values=[]
    for value in outer_top:
        filt= (df[column_name] == value )
        top_values.extend([x for x in df[filt][Column_to_count].value_counts(ascending= False).head(no_of_top_topick).index])
    return top_values

# **Top Labels and their encoding**

## **Top 10 primary description**

In [152]:
top_prim_des= [x for x in df["primary_description"].value_counts(ascending= False).head(10).index]
top_values = top_prim_des.copy()
top_values

['THEFT',
 'BATTERY',
 'CRIMINAL DAMAGE',
 'ASSAULT',
 'DECEPTIVE PRACTICE',
 'OTHER OFFENSE',
 'NARCOTICS',
 'BURGLARY',
 'MOTOR VEHICLE THEFT',
 'ROBBERY']

### **Encoding Top Primary Description values**

In [153]:
column_name= "primary_description"
encoder_function(column_name, top_values)

In [154]:
df.head(2)

block  ... primary_description_ROBBERY
0   061XX W GRAND AVE  ...                           0
1  043XX N CICERO AVE  ...                           0

[2 rows x 25 columns]

## **Top Secondary crimes for the top primary description**

In [155]:
# to pass
column_name= "primary_description" # The name of the outer level column
outer_top= top_prim_des # Top values of outer column
no_of_top_topick =10 #these are how many number of values to pick for the inner top values
Column_to_count= "secondary_description" # Name of column from which you want to pick top values of

top_values= list(set(list_inner_top_values(column_name, outer_top, no_of_top_topick, Column_to_count)))

print(top_values)
print(len(set(top_values)))

['POSS: CANNABIS MORE THAN 30GMS', 'TO VEHICLE', 'HARASSMENT BY ELECTRONIC MEANS', 'FOUND SUSPECT NARCOTICS', 'VIOLATE ORDER OF PROTECTION', 'PURSE-SNATCHING', 'THEFT / RECOVERY - AUTOMOBILE', 'AGGRAVATED: OTHER DANG WEAPON', 'OTHER VEHICLE OFFENSE', 'VEHICLE TITLE/REG OFFENSE', 'ATTEMPT FORCIBLE ENTRY', 'TELEPHONE THREAT', 'INSTITUTIONAL VANDALISM', 'ATTEMPT: STRONGARM-NO WEAPON', 'AGGRAVATED DOMESTIC BATTERY: HANDS/FIST/FEET SERIOUS INJURY', 'ATTEMPT THEFT', 'GUN OFFENDER: ANNUAL REGISTRATION', 'HOME INVASION', 'STRONGARM - NO WEAPON', 'ARMED - HANDGUN', 'MANU/DEL:CANNABIS OVER 10 GMS', 'DELIVERY CONTAINER THEFT', 'FALSE/STOLEN/ALTERED TRP', 'TO CITY OF CHICAGO PROPERTY', 'UNLAWFUL ENTRY', 'POCKET-PICKING', 'VEHICULAR HIJACKING', 'RETAIL THEFT', 'AGGRAVATED DOMESTIC BATTERY: OTHER DANG WEAPON', 'DOMESTIC BATTERY SIMPLE', 'ARMED: HANDGUN', 'AGGRAVATED VEHICULAR HIJACKING', 'FINANCIAL IDENTITY THEFT OVER $ 300', 'TO PROPERTY', 'FORGERY', 'OTHER CRIME AGAINST PERSON', 'FORCIBLE ENTRY', 

#### **Encoding Top secondary crimes**

In [156]:
column_name= "secondary_description"
encoder_function(column_name, top_values)

In [157]:
df.head(2)

block  ... secondary_description_THEFT/RECOVERY: TRUCK,BUS,MHOME
0   061XX W GRAND AVE  ...                                                  0   
1  043XX N CICERO AVE  ...                                                  0   

[2 rows x 111 columns]

## **Top iucr codes for the top primary description**

In [158]:
# to pass
column_name= "primary_description" # The name of the outer level column
outer_top= top_prim_des # Top values of outer column
no_of_top_topick =10 #these are how many number of values to pick for the inner top values
Column_to_count= "iucr" # Name of column from which you want to pick top values of

top_values= list(set(list_inner_top_values(column_name, outer_top, no_of_top_topick, Column_to_count)))

In [159]:
print(top_values)
print(len(set(top_values)))

['0420', '0880', '033A', '0313', '1811', '1812', '1822', '0918', '1370', '0454', '1305', '0497', '0530', '0520', '502R', '0325', '1150', '5111', '0486', '0430', '1154', '0920', '0895', '0925', '502P', '0850', '1122', '0496', '0870', '0326', '0650', '0865', '0553', '0890', '0935', '0860', '0545', '1320', '2024', '0340', '051B', '4387', '2027', '1156', '1152', '0557', '051A', '0498', '0820', '1345', '1121', '2826', '1310', '0910', '0610', '0937', '0320', '2014', '0312', '0810', '0917', '0330', '0630', '0558', '1340', '1130', '031A', '1375', '2022', '0560', '0460', '0620', '5002', '0927', '1153', '0930', '0915', '0484', '5000', '2028', '1120', '0334', '0554', '2093', '1210', '4650', '2820', '2092', '2825', '041A']
90


#### **Encoding Top IUCR codes**

In [160]:
column_name= "iucr"
encoder_function(column_name, top_values)

In [161]:
df.head(2)

block  iucr  ... iucr_2825 iucr_041A
0   061XX W GRAND AVE  0486  ...         0         0
1  043XX N CICERO AVE  0560  ...         0         0

[2 rows x 201 columns]

## **Top Blocks from all wards**

In [162]:
# Getting all blocks
wards= list(df["ward"].unique())
print(wards)

[36.0, 45.0, 28.0, 21.0, 42.0, 39.0, 1.0, 17.0, 8.0, 41.0, 27.0, 32.0, 9.0, 7.0, 48.0, 24.0, 25.0, 22.0, 47.0, 29.0, 26.0, 31.0, 40.0, 49.0, 10.0, 38.0, 33.0, 4.0, 20.0, 18.0, 44.0, 50.0, 5.0, 3.0, 14.0, 37.0, 43.0, 6.0, 34.0, 12.0, 2.0, 16.0, 35.0, 46.0, 15.0, 23.0, 19.0, 11.0, 30.0, 13.0]


In [163]:
# Getting list of blocks that occur in top wards
# to pass
column_name= "ward" # The name of the outer level column
outer_top= wards # Top values of outer column
no_of_top_topick =10 #these are how many number of values to pick for the inner top values
Column_to_count= "block" # Name of column from which you want to pick top values of

top_values= list(set(list_inner_top_values(column_name, outer_top, no_of_top_topick, Column_to_count)))
print(top_values)

['003XX E RANDOLPH ST', '089XX S COMMERCIAL AVE', '016XX W HOWARD ST', '080XX S WESTERN AVE', '010XX N FRANCISCO AVE', '001XX N CENTRAL AVE', '035XX N BROADWAY', '048XX S WOOD ST', '005XX W TAYLOR ST', '005XX N MICHIGAN AVE', '076XX S CICERO AVE', '016XX N WELLS ST', '040XX W LAWRENCE AVE', '046XX W DIVERSEY AVE', '026XX N NARRAGANSETT AVE', '011XX W 63RD ST', '028XX W CERMAK RD', '035XX S RHODES AVE', '004XX E 79TH ST', '087XX S COTTAGE GROVE AVE', '056XX W MADISON ST', '020XX W BALMORAL AVE', '017XX E 67TH ST', '0000X E RANDOLPH ST', '056XX W FULLERTON AVE', '132XX S LANGLEY AVE', '037XX N BROADWAY', '061XX S ASHLAND AVE', '082XX S DREXEL AVE', '003XX W 64TH ST', '034XX W FULLERTON AVE', '094XX S ASHLAND AVE', '065XX S PULASKI RD', '0000X W 95TH ST', '006XX E 131ST ST', '035XX W ARMITAGE AVE', '007XX E 79TH ST', '011XX S CANAL ST', '020XX N HALSTED ST', '064XX N SHERIDAN RD', '013XX S CANAL ST', '060XX W 63RD ST', '024XX W CHICAGO AVE', '035XX S MICHIGAN AVE', '076XX S HALSTED ST', '

### **Encoding Block from wards**

In [164]:
column_name= "block"
encoder_function(column_name, top_values)

In [165]:
df.head(2)

block  ... block_034XX N WESTERN AVE
0   061XX W GRAND AVE  ...                         0
1  043XX N CICERO AVE  ...                         0

[2 rows x 698 columns]

## **Top beats from top wards**

In [166]:
# Getting list of blocks that occur in top wards
# to pass
column_name= "ward" # The name of the outer level column
outer_top= wards # Top values of outer column
no_of_top_topick =1 #these are how many number of values to pick for the inner top values
Column_to_count= "beat" # Name of column from which you want to pick top values of

top_values= list(set(list_inner_top_values(column_name, outer_top, no_of_top_topick, Column_to_count)))
print(top_values)

[2433, 131, 1924, 1033, 524, 1911, 1424, 1935, 913, 533, 1432, 922, 923, 2212, 421, 1833, 1834, 813, 1712, 432, 823, 312, 1723, 1724, 835, 1731, 713, 331, 2512, 1623, 1113, 2521, 2011, 1631, 1121, 2533, 2535, 2412, 621, 622, 623, 1522, 1651, 1011, 1013, 2422, 631, 1914, 123, 124]


#### **Encoding Top Beats**

In [167]:
column_name= "beat"
top_values= list(map(str,top_values))
encoder_function(column_name, top_values)

In [168]:
df.head(2)

block  iucr primary_description  ... beat_1914 beat_123 beat_124
0   061XX W GRAND AVE  0486             BATTERY  ...         0        0        0
1  043XX N CICERO AVE  0560             ASSAULT  ...         0        0        0

[2 rows x 748 columns]

## **Top Locations**

In [169]:
top_prim_des= [x for x in df["location_description"].value_counts(ascending= False).head(30).index]
top_values = top_prim_des.copy()
top_values

['STREET',
 'RESIDENCE',
 'APARTMENT',
 'SIDEWALK',
 'OTHER',
 'SMALL RETAIL STORE',
 'RESTAURANT',
 'PARKING LOT/GARAGE(NON.RESID.)',
 'ALLEY',
 'DEPARTMENT STORE',
 'VEHICLE NON-COMMERCIAL',
 'GAS STATION',
 'RESIDENCE PORCH/HALLWAY',
 'RESIDENTIAL YARD (FRONT/BACK)',
 'GROCERY FOOD STORE',
 'RESIDENCE-GARAGE',
 'CONVENIENCE STORE',
 'OTHER (SPECIFY)',
 'CTA TRAIN',
 'SCHOOL, PUBLIC, BUILDING',
 'BAR OR TAVERN',
 'COMMERCIAL / BUSINESS OFFICE',
 'PARK PROPERTY',
 'PARKING LOT / GARAGE (NON RESIDENTIAL)',
 'DRUG STORE',
 'RESIDENCE - PORCH / HALLWAY',
 'RESIDENCE - YARD (FRONT / BACK)',
 'CTA BUS',
 'CTA STATION',
 'BANK']

### **Encoding Top Locations**

In [170]:
column_name= "location_description"
encoder_function(column_name, top_values)

In [171]:
df.head(2)

block  ... location_description_BANK
0   061XX W GRAND AVE  ...                         0
1  043XX N CICERO AVE  ...                         0

[2 rows x 778 columns]

## **Top FBI Code**

In [172]:
top_prim_des= [x for x in df["fbi_cd"].value_counts(ascending= False).head(10).index]
top_values = top_prim_des.copy()
top_values

['06', '08B', '14', '26', '08A', '11', '18', '05', '07', '03']

### **Encoding top FBI codes**

In [173]:
column_name= "fbi_cd"
encoder_function(column_name, top_values)

In [174]:
df.head(2)

block  iucr  ... fbi_cd_07 fbi_cd_03
0   061XX W GRAND AVE  0486  ...         0         0
1  043XX N CICERO AVE  0560  ...         0         0

[2 rows x 788 columns]

# **Label Encoding of other  columns**

## **Encoding Arrest Column**

In [175]:
column_name= "arrest"

labels = list(df[column_name].unique())
encoder_function(column_name, labels)
df.head(2)

block  iucr primary_description  ... fbi_cd_03 arrest_N arrest_Y
0   061XX W GRAND AVE  0486             BATTERY  ...         0        1        0
1  043XX N CICERO AVE  0560             ASSAULT  ...         0        0        1

[2 rows x 790 columns]

## **Encoding domestic Column**

In [176]:
column_name= "domestic"

labels = list(df[column_name].unique())
encoder_function(column_name, labels)
df.head(2)

block  iucr  ... domestic_N domestic_Y
0   061XX W GRAND AVE  0486  ...          1          0
1  043XX N CICERO AVE  0560  ...          1          0

[2 rows x 792 columns]

## **Encoding domestic Column**

In [177]:
column_name= "domestic"

labels = list(df[column_name].unique())
encoder_function(column_name, labels)
df.head(2)

block  iucr  ... domestic_N domestic_Y
0   061XX W GRAND AVE  0486  ...          1          0
1  043XX N CICERO AVE  0560  ...          1          0

[2 rows x 792 columns]

## **Encoding month Column**

In [178]:
column_name= "month"

labels = list(df[column_name].unique())
encoder_function(column_name, labels)
df.head(2)

block  iucr  ... month_Apr month_May
0   061XX W GRAND AVE  0486  ...         0         0
1  043XX N CICERO AVE  0560  ...         0         0

[2 rows x 804 columns]

## **Encoding weekday Column**

In [179]:
column_name= "week_day"

labels = list(df[column_name].unique())
encoder_function(column_name, labels)
df.head(2)

block  iucr  ... week_day_Tuesday week_day_Wednesday
0   061XX W GRAND AVE  0486  ...                0                  0
1  043XX N CICERO AVE  0560  ...                0                  0

[2 rows x 811 columns]

## **Encoding hour Column**

In [180]:
column_name= "hour"

labels = list(df[column_name].unique())

labels= list(map(str,labels))
encoder_function(column_name, labels)
df.head(2)

block  iucr primary_description  ... hour_2 hour_3 hour_4
0   061XX W GRAND AVE  0486             BATTERY  ...      0      0      0
1  043XX N CICERO AVE  0560             ASSAULT  ...      0      0      0

[2 rows x 835 columns]

# **Scaling Values**

## **Scaling longitude Column**

In [181]:
df["latitude"].head(5)

0    41.921325
1    41.960495
2    41.866332
3    41.740986
4    41.893282
Name: latitude, dtype: float64

## **Scaling latitude Column**

In [182]:
df["latitude"].head(5)

0    41.921325
1    41.960495
2    41.866332
3    41.740986
4    41.893282
Name: latitude, dtype: float64

# **Deleting Label Columns**

In [183]:
data.columns

Index(['block', 'iucr', 'primary_description', 'secondary_description',
       'location_description', 'arrest', 'domestic', 'beat', 'ward', 'fbi_cd',
       'latitude', 'longitude', 'month', 'week_day', 'hour'],
      dtype='object')

In [184]:
del df["block"]
del df["iucr"]
del df["primary_description"]
del df["secondary_description"]
del df["location_description"]
del df["arrest"]
del df["domestic"]
del df["beat"]
del df["fbi_cd"]
del df["month"]
del df["week_day"]
del df["hour"]

In [185]:
df.head(2)

ward   latitude  longitude  ...  hour_2  hour_3  hour_4
0   36  41.921325 -87.779766  ...       0       0       0
1   45  41.960495 -87.747514  ...       0       0       0

[2 rows x 823 columns]

# **Saving Ward data**

In [186]:
df.to_csv("ward_data.csv")